No completion before the first character or in an empty line

pghvlaans opened the pull request:
There are still two contexts in which the unhelpful "alias + builtin + path" completion occurs:

* After the start of an empty line
* After the start of a line, but before the first character

To stop completion from being attempted in these contexts, a loop is added to `vi.c` and `emacs.c` that looks over the previous positions (current position inclusive) and allows the completion process to continue once it finds the first non-space character. The existing placement restrictions remain in case the cursor arrives in a negative position via a bug.

pghvlaans committed:
No completion before the first character or in an empty line

McDutchie commented on src/cmd/ksh93/edit/emacs.c:
```diff
@@ -1002,7 +1004,13 @@ static int escape(register Emacs_t* ep,register genchar *out,int count)
 			ep->mark = cur;
-			if(cur<1)
+			for(x=0; x <= cur; x++)
+				if(out[x] > ' ')
```
This is ASCII-centric. This code base still theoretically supports EBCDIC (though I've been unable to test it). Also, not all control characters are whitespace. Better would be: `if(!isspace(out[x]))`

McDutchie commented on src/cmd/ksh93/edit/vi.c:
```diff
@@ -2545,7 +2547,13 @@ static int textmod(register Vi_t *vp,register int c, int mode)
 	case '=':		/** list file name expansions **/
-		if( cur_virt == INVALID )
+		for(x=0; x <= cur_virt; x++)
+			if(virtual[x] > ' ')
```
Here, too, I would prefer: `if(!isspace(virtual[x]))`

pghvlaans commented:
Sorry, I pushed the wrong button. Testing the changes now.

pghvlaans committed:
Use isspace in preference to ASCII values

McDutchie commented:
Hmm. The changes I suggested don't actually seem to work. :(  I'm not yet clear why that is.

McDutchie commented:
Ah, it's because I failed to consider the terminating 0 byte. The fix is to avoid including it in the loop by using `<` instead of `<=`.
```diff
diff --git a/src/cmd/ksh93/edit/emacs.c b/src/cmd/ksh93/edit/emacs.c
index 2b0f6771b..d70079524 100644
--- a/src/cmd/ksh93/edit/emacs.c
+++ b/src/cmd/ksh93/edit/emacs.c
@@ -1004,12 +1004,14 @@ static int escape(register Emacs_t* ep,register genchar *out,int count)
 		case '*':		/* filename expansion */
 		case '=':	/* escape = - list all matching file names */
 			ep->mark = cur;
-			for(x=0; x <= cur; x++)
+			for(x=0; x < cur; x++)
+			{
 				if(!isspace(out[x]))
 				{
 					allempty = 0;
 					break;
 				}
+			}
 			if(cur<1 || allempty==1)
 			{
 				beep();
diff --git a/src/cmd/ksh93/edit/vi.c b/src/cmd/ksh93/edit/vi.c
index 6c7996bed..6ed8028e5 100644
--- a/src/cmd/ksh93/edit/vi.c
+++ b/src/cmd/ksh93/edit/vi.c
@@ -2547,12 +2547,14 @@ addin:
 	case '*':		/** do file name expansion in place **/
 	case '\\':		/** do file name completion in place **/
 	case '=':		/** list file name expansions **/
-		for(x=0; x <= cur_virt; x++)
+		for(x=0; x < cur_virt; x++)
+		{
 			if(!isspace(virtual[x]))
 			{
 				allempty = 0;
 				break;
 			}
+		}
 		if( cur_virt == INVALID || allempty == 1 )
 			return(BAD);
 		/* FALLTHROUGH */
```

pghvlaans commented:
`emacs` mode needs `<`; it fixes the issue in Linux and OpenBSD. `vi.c` still works with`<=` `<` causes it to fail to complete from the second position. `cur_virt` and `out` seem to be set up a little differently.

pghvlaans committed:
Use <= in emacs mode

McDutchie commented:
There is another problem in vi mode. If you press tab after some blanks in input mode, it leaves input mode and returns to command mode. With tab, it should stay in input mode.

pghvlaans committed:
vi: Keep insert mode in case of failed completion due to empty line

pghvlaans commented:
The most recent commit should do the trick. Tested on Linux, MacOS and some BSDs.

McDutchie committed:
Pedantic tweaks

- Use more local scope for allempty and x variables
- Use char for boolean-like allempty variable, omit ==1
- Use braces to include another braces block

pghvlaans commented:
"Pedantic" can be pretty darn important so far as shells are concerned. No quibble. Thanks!
